howso/boundary_values.amlg

```diff
@@ -0,0 +1,879 @@
+											)
+										)
+									)
+
+								)
+								boundary_action_condition
+							)
+						))
+					)
+
+					;no specified boundary, define a function that will indicate a boundary based on action values
+					(let
+						(assoc
+							num_nominal_features (size (filter (lambda (contains_index !nominalsMap (current_value))) action_features))
+						)
+						(if (= (size action_features) num_nominal_features)
+							;if just nominals, any change will indicate a boundary crossed
+							(lambda (!= action_values new_action_values))
+
+							;else there is some non-nominal thing, so do the ratio of total distance to context distance
+							(lambda
+								(/
+									;difference from original context/action to updated context/action
+									(generalized_distance
+										feature_weights
+										!queryDistanceTypeMap
+										query_feature_attributes_map
+										feature_deviations
+										p_parameter
+										(append context_values action_values)
+										(append new_context_values new_action_values)
+										(append context_features action_features)
+										(null)
+									)
+									;difference from original context to updated context
+									(generalized_distance
+										feature_weights
+										!queryDistanceTypeMap
+										query_feature_attributes_map
+										feature_deviations
+										p_parameter
+										context_values
+										new_context_values
+										context_features
+										(null)
+									)
+								)
+							)
+						)
+					)
+				)
+		))
+
+		;if well_defined_boundary is true, then eval_function returns a boolean indicating if a/the boundary has
+		;been crossed. If false, eval_function returns a ratio of the difference in context/action to difference in context
+		;and the boundary_values returned will attempt to maximize this ratio.
+		(declare (assoc well_defined_boundary (contains_value ["!=" "apply"] (get_type_string eval_function)) ))
+
+		(declare (assoc
+			boundary_values
+				;flow where there must just be some difference in the action_value
+				(map
+					(lambda
+						(let
+							(assoc
+								query_feature (current_index 1)
+								query_feature_index (get context_feature_index_map (current_index 1))
+								remaining_context_features (filter (lambda (!= (current_index 2) (current_value))) context_features)
+								feature_value (get context_values (get context_feature_index_map (current_index 1)))
+							)
+
+							(if (contains_index !nominalsMap query_feature)
+								;Nominals Block
+								(call !ComputeNominalBoundaryValues)
+
+								(contains_value !ordinalFeatures query_feature)
+								;Ordinals Block
+								(call !ComputeOrdinalBoundaryValues)
+
+
+								(contains_index !editDistanceFeatureTypesMap query_feature)
+								;Not supported.
+								(null)
+
+								;else will assume continuous... TODO: this probably isn't exactly correct, the not of all previous conditions might not imply its continuous
```

Comment:
either need ticket or remove todo

Reply:
I think this is probably correct now. Put this comment in before other blocks were added